onnx/test/cpp/function_verify_test.cc

```diff
@@ -0,0 +1,151 @@
+#include <iostream>
+#include <set>
+#include "gtest/gtest.h"
+#include "onnx/checker.h"
+#include "onnx/common/constants.h"
+#include "onnx/defs/schema.h"
+#include "onnx/onnx-operators_pb.h"
+#include "onnx/onnx_pb.h"
+
+namespace ONNX_NAMESPACE {
+namespace Test {
+using namespace checker;
+using TENSOR_TYPES_MAP =
+    std::unordered_map<std::string, std::vector<std::string>>;
+
+void VerifyTypeConstraint(
+    const OpSchema& function_op,
+    const FunctionProto* function_proto) {
+  // TC for function nodes should satisfy the definition defined in the opschema
+  // This is designed to be a best-effort test
+  TENSOR_TYPES_MAP tc_map;
+  std::set<std::string> primitive_types(
+      OpSchema::all_numeric_types().begin(),
+      OpSchema::all_numeric_types().end());
+  for (const auto& input : function_op.inputs()) {
+    std::string name = input.GetName();
+    for (const auto& t : input.GetTypes()) {
+      if (!primitive_types.count(*t)) {
+        return; // skip variable types check for now
+      }
+      tc_map[name].emplace_back(*t);
+    }
+  }
+
+  for (const auto& output : function_op.outputs()) {
+    std::string name = output.GetName();
+    for (const auto& t : output.GetTypes()) {
+      if (!primitive_types.count(*t)) {
+        return; // skip variable types check for now
+      }
+      tc_map[name].emplace_back(*t);
+    }
+  }
+
+  for (auto& node : function_proto->node()) {
+    std::string op_type = node.op_type();
+    const OpSchema* schema = OpSchemaRegistry::Schema(
+        op_type, function_op.since_version(), function_op.domain());
+
+    std::unordered_map<std::string, int> input_tensor_name_idx_map;
+    std::unordered_map<std::string, int> output_tensor_name_idx_map;
+    // Enforce it on input
+    for (unsigned int i = 0; i < schema->inputs().size(); ++i) {
+      auto& input = schema->inputs().at(i);
+      input_tensor_name_idx_map[input.GetName()] = i;
+    }
+    for (auto& tensor_name_tc : tc_map) {
+      auto iter = input_tensor_name_idx_map.find(tensor_name_tc.first);
+      if (iter == input_tensor_name_idx_map.end())
+        continue;
+      const auto& types = schema->inputs().at(iter->second).GetTypes();
+      std::unordered_set<std::string> allowed_types;
+      for (auto& s : types) {
+        allowed_types.insert(*s);
+      }
+      for (auto& type : tensor_name_tc.second) {
+        if (allowed_types.find(type) == allowed_types.end()) {
+          fail_check(
```

Comment:
We can't call this an error yet, because the elements in allowed_types might be a type variable. Furthermore, the two separate loops (lines 53 to 56 and the one in line 57 is not required, since we should be able to directly do the check at line 55 for the ith argument.

Reply:
Hi Rama.. sorry just saw your comment. see L22-L29. I exclude the type variable already. This one is only enforced on primitive types

Reply:
For the other two comments: the check_function is a existing function. I didnt change any of that (just misplaced it so there is diff.) Ill put it back and see if we can do any modification on that. For VerifyFunctionNode Ill remove it. Thanks for pointing that out 😃

Reply:
L23 should be all_tensor_types(). Ill change it.

Reply:
Hi Raymond, you excluded it in tc_map. But allowed_types can include type-variables as well. You have not excluded it from that. In any case, I don't think that the overall structure for this type-checking is the right one for extending it to handle all cases.

Reply:
Sorry I am not familiar with the type variable: If a node in a function node, which consumes or produce a tensor with variable type, and is also a exposed tensor of the whole function op, should this variable type also appear in tc_map, which generated from the overall defined type constraints?

Reply:
I try to figure out a way to check constaints between op's type variable and node inside function body but somehow cannot find a specific rule in order to generate one.. I think we can explore how to resolve this later.. its a huge item

Reply:
Sorry, ignore my comment about type-variables. I see that "Finalize" on OpSchema would have converted the type-variable to a list of proper types using type-constraints already. (But I still think we will need to replace this with the logic we have in "Graph::InferAndVerifyTypeMatch(…)" etc.)

Reply:
But we can do that separately. Can you just add a comment line saying "TODO: apply complete type inference across function body for a more complete check?"

Reply:
Sure. I already put a todo there.